src/librustc/middle/resolve_lifetime.rs

```diff
@@ -1087,24 +1088,25 @@ fn check_mixed_explicit_and_in_band_defs(
     tcx: TyCtxt<'_, '_, '_>,
     params: &P<[hir::GenericParam]>,
 ) {
-    let lifetime_params: Vec<_> = params.iter().filter_map(|param| match param.kind {
+    let mut lifetime_params = params.iter().filter_map(|param| match param.kind {
         GenericParamKind::Lifetime { kind, .. } => Some((kind, param.span)),
         _ => None,
-    }).collect();
-    let explicit = lifetime_params.iter().find(|(kind, _)| *kind == LifetimeParamKind::Explicit);
-    let in_band = lifetime_params.iter().find(|(kind, _)| *kind == LifetimeParamKind::InBand);
+    });
+
+    let explicit = lifetime_params.by_ref().find(|(kind, _)| *kind == LifetimeParamKind::Explicit);
+    let in_band = lifetime_params.find(|(kind, _)| *kind == LifetimeParamKind::InBand);
```

Comment:
This looks a bit problematic. From the code it's not entirely clear if the second `find` will see any elements, since they might have been consumed by the first `find`. I'd just leave this as is.